### PR TITLE
WI 969 Forbid Sync Clause for Strict Types

### LINKED ISSUE #960: Forbird SYNC clause with Strict Type
It seems that Cobol 2002 type comparison is more complicated that what we implemented:

> Two typed items are of the same type when:
> - The items are described with TYPE clauses that reference equivalent type declarations; or
> - The items are described as subordinate items in equivalent type declarations, starting at the same relative byte position and having the same length in bytes.
> 
> 
> Two type declarations are considered equivalent when they have the same type-name, and for each 
> elementary item in one type declaration there is a corresponding elementary item in the other type 
> declaration, starting at the same relative byte position and having the same length in bytes. 
> Each pair of corresponding elementary items shall have the same BLANK WHEN ZERO, DYNAMIC LENGTH, JUSTIFIED, PICTURE, SIGN, SYNCHRONIZED, and USAGE clauses, with the following exceptions:
> 1) Currency symbols match if and only if the corresponding currency strings are the same.
> 2) Period picture symbols match if and only if the DECIMAL-POINT IS COMMA clause is in effect for both or for
> neither of these type declarations. Comma picture symbols match if and only if the DECIMAL-POINT IS
> COMMA clause is in effect for both or for neither of these type declarations.
> 
> Additionally, locale specifications in the PICTURE clauses match if and only if:
> - both specify the same SIZE phrase in the LOCALE phrase of the PICTURE clause, and
> - both specify the LOCALE phrase without a locale-name or both specify the LOCALE phrase with the same
> external identification, where the external identification is the external-locale-name or literal value
> associated with a locale-name in the LOCALE clause of the SPECIAL-NAMES paragraph.
> 

> NOTE:  When two typed items are defined in the same source element, the above rules mean that either both items are described with the same TYPE clause or both items are described as the same subordinate item in the same type declaration.

As we only check type name:
For Scrict type which is our own definition we need to forbid Sync clause to ensure type really match at runtime.

- [ ] Forbid SYNC in a Typedef
   -Already done at the typedef level (d1686a4) but not for data under the typedef.
- [ ] A strictly typed variable cannot be SYNC
  - SYNC can be define at a parent level.

### PR DESCRIPTION
**Pull request by rooksdo**
Closes #960

### CODE REVIEW

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 1]
```diff
@@ -389,6 +389,49 @@ public override bool Visit(DataDefinition dataDefinition)
...
+                //Variable inside Typedef => check if parent typedef is marked Strict
+                else if (dataDefinition.SemanticData.NearestParent(Symbol.Kinds.Typedef).HasFlag(Symbol.Flags.Strict))
+                {
+                    DiagnosticUtils.AddError(dataDefinition, $"{dataDefinition.Name} cannot be declared with Sync clause because it is part of Strict Type definition {dataDefinition.ParentTypeDefinition?.Name}.");
+                }
+
+            }
+            
+            //Variable is of a user defined Strict Type
+            if (dataDefinition.SemanticData.HasFlag(Symbol.Flags.HasATypedefType) && dataDefinition.DataType.RestrictionLevel == RestrictionLevel.STRICT)
```
Comment (smedilol) [inline TypeCobol/Compiler/Diagnostics/CrossChecker.cs:410]: DataType is the old type mechanism that must be replaced by new Type class.
New Type class doesn't have the resolved TypeDefinition yet.
You can use dataDefinition.TypeDefinition?.RestrictionLevel in the meantime.

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 2]
```diff
@@ -389,6 +389,49 @@ public override bool Visit(DataDefinition dataDefinition)
...
+            {
+                //Sync clause not allowed on variable declaration because its type is Strict
+                if (dataDefinition.SemanticData.HasFlag(Symbol.Flags.Sync))
+                {
+                    DiagnosticUtils.AddError(dataDefinition, $"{dataDefinition.Name} cannot be declared with Sync clause because its Type definition {dataDefinition.DataType.Name} is Strict.");
+                }
+                //Check whether variable is under a sync parent
+                else if (dataDefinition.CodeElement?.LevelNumber?.Value > 1)
```
Comment (fm-117) [inline TypeCobol/Compiler/Diagnostics/CrossChecker.cs:418]: dataDefinition.SemanticData can be converted to VariableSymbol which has a Level property. Start by casting dataDefinition.SemanticData so you can use it all the way in the checker.

File: TypeCobol/Compiler/Diagnostics/CrossChecker.cs [hunk 3]
```diff
@@ -389,6 +389,49 @@ public override bool Visit(DataDefinition dataDefinition)
...
+                {
+                    //Retrieve higher parent
+                    Symbol parent = dataDefinition.SemanticData.Owner;
```
Comment (smedilol) [inline TypeCobol/Compiler/Diagnostics/CrossChecker.cs:421]: I thought that sync flag were inherited.
Anyway, can you replace this code by an inheritance of Sync flag?
This way the `if (dataDefinition.SemanticData.HasFlag(Symbol.Flags.Sync))` will be enough to test this rule.

Comment (fm-117) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs:421]: Could you update your branch with latest commits from 1721_SemanticDomainLite please ? 558b8e75295dd0dab6d8c8fcd330bdc4e8b194f3 should give you inherited SYNC flags.

Comment (rooksdo) [reply TypeCobol/Compiler/Diagnostics/CrossChecker.cs:421]: Ok now sync flag is inherited, but the information that it is inherited is not present. So those two cases are indistinguishable using only Symbol class:
```cob
01 MyGroup sync.
    05 MyTypeVar type MyTypeStrict.
```
and
```cob
01 MyGroup.
    05 MyTypeVar type MyTypeStrict sync.
```

**Comment by smedilol** [comment]
Beware, there is a mix of issue ID in commit : 960 and 969.
The char `#` is also missing before issue ID in commit.

**Comment by rooksdo** [comment]
See rebase PR #1733

**Pull request closed without merging**